File: jsrc/legacy/SxxMachine/builtin/PRED_loggable_1.java
package com.googlecode.prolog_cafe.builtin;

import java.util.HashMap;
import java.util.Map;
import java.util.logging.Level;
import java.util.logging.Logger;

import com.googlecode.prolog_cafe.exceptions.IllegalTypeException;
import com.googlecode.prolog_cafe.exceptions.PrologException;
import com.googlecode.prolog_cafe.lang.Operation;
import com.googlecode.prolog_cafe.lang.Predicate.P1;
import com.googlecode.prolog_cafe.lang.Prolog;
import com.googlecode.prolog_cafe.lang.StructureTerm;
import com.googlecode.prolog_cafe.lang.SymbolTerm;
import com.googlecode.prolog_cafe.lang.Term;

/**
 * <p><b>loggable(package:level)</b> - succeed if given level is loggable for given package.
 * If package is absent, then current package is automatically added by prolog compiler.
 * So call loggable('INFO') is valid. Both package and level are expected to be atoms. (Variables will cause errors).
 * Level can be one of 'OFF','SEVERE','WARNING','INFO','CONFIG','FINE','FINER','FINEST','ALL' (from highest to lowest).
 * Also lower case variants without quotes are accepted.  
 *
 * <p>The predicate finds {@link Logger} instance, corresponding to given package, and calls its {@link Logger#isLoggable(Level)}
 * method with given level.
 * 	  
 * @author semenov
 *
 */
public class PRED_loggable_1 extends P1 {

	static final Map<Term, Level> LEVELS = new HashMap<>();
	static {
		LEVELS.put(SymbolTerm.intern("ALL"), Level.ALL);
		LEVELS.put(SymbolTerm.intern("all"), Level.ALL);
		LEVELS.put(SymbolTerm.intern("FINEST"), Level.FINEST);
		LEVELS.put(SymbolTerm.intern("finest"), Level.FINEST);
		LEVELS.put(SymbolTerm.intern("FINER"), Level.FINER);
		LEVELS.put(SymbolTerm.intern("finer"), Level.FINER);
		LEVELS.put(SymbolTerm.intern("FINE"), Level.FINE);
		LEVELS.put(SymbolTerm.intern("fine"), Level.FINE);
		LEVELS.put(SymbolTerm.intern("INFO"), Level.INFO);
		LEVELS.put(SymbolTerm.intern("info"), Level.INFO);
		LEVELS.put(SymbolTerm.intern("CONFIG"), Level.CONFIG);
		LEVELS.put(SymbolTerm.intern("config"), Level.CONFIG);
		LEVELS.put(SymbolTerm.intern("WARNING"), Level.WARNING);
		LEVELS.put(SymbolTerm.intern("warning"), Level.WARNING);
		LEVELS.put(SymbolTerm.intern("SEVERE"), Level.SEVERE);
		LEVELS.put(SymbolTerm.intern("severe"), Level.SEVERE);
	}

	public PRED_loggable_1(Term arg1, Operation cont) {
		this.arg1 = arg1;
		this.cont = cont;
	}
	
	@Override
	public Operation exec(Prolog engine) throws PrologException {

		Term a1 = arg1.dereference();

		if (!(a1 instanceof StructureTerm) || a1.arity()!=2){
			throw new IllegalTypeException(this, 1, "package:level", a1);
		}
		
		Logger logger = Logger.getLogger(a1.arg(0).name());
		if (!logger.isLoggable(LEVELS.getOrDefault(a1.arg(1), Level.INFO))){
			return engine.fail();
		}
		return cont;
	}

}

File: src/legacy/SxxMachine/builtin/PRED_loggable_1.cpp
using namespace std;

#include "PRED_loggable_1.h"
#include "../../../machine/SxxMachine/Term.h"
#include "../../../machine/SxxMachine/SymbolTerm.h"
#include "../../../machine/SxxMachine/Operation.h"
#include "../../../machine/SxxMachine/Prolog.h"
#include "../../../exceptions/SxxMachine/PrologException.h"
#include "../../../machine/SxxMachine/StructureTerm.h"
#include "../../../exceptions/SxxMachine/IllegalTypeException.h"

namespace SxxMachine
{
	using IllegalTypeException = SxxMachine::IllegalTypeException;
	using PrologException = SxxMachine::PrologException;
	using Operation = SxxMachine::Operation;
	using P1 = SxxMachine::Predicate::P1;
	using Prolog = SxxMachine::Prolog;
	using StructureTerm = SxxMachine::StructureTerm;
	using SymbolTerm = SxxMachine::SymbolTerm;
	using Term = SxxMachine::Term;
const unordered_map<SxxMachine::Term*, java::util::logging::Level*> PRED_loggable_1::LEVELS = unordered_map<SxxMachine::Term*, java::util::logging::Level*>();

	PRED_loggable_1::StaticConstructor::StaticConstructor()
	{
	LEVELS.emplace(SymbolTerm::intern("ALL"), Level::ALL);
	LEVELS.emplace(SymbolTerm::intern("all"), Level::ALL);
	LEVELS.emplace(SymbolTerm::intern("FINEST"), Level::FINEST);
	LEVELS.emplace(SymbolTerm::intern("finest"), Level::FINEST);
	LEVELS.emplace(SymbolTerm::intern("FINER"), Level::FINER);
	LEVELS.emplace(SymbolTerm::intern("finer"), Level::FINER);
	LEVELS.emplace(SymbolTerm::intern("FINE"), Level::FINE);
	LEVELS.emplace(SymbolTerm::intern("fine"), Level::FINE);
	LEVELS.emplace(SymbolTerm::intern("INFO"), Level::INFO);
	LEVELS.emplace(SymbolTerm::intern("info"), Level::INFO);
	LEVELS.emplace(SymbolTerm::intern("CONFIG"), Level::CONFIG);
	LEVELS.emplace(SymbolTerm::intern("config"), Level::CONFIG);
	LEVELS.emplace(SymbolTerm::intern("WARNING"), Level::WARNING);
	LEVELS.emplace(SymbolTerm::intern("warning"), Level::WARNING);
	LEVELS.emplace(SymbolTerm::intern("SEVERE"), Level::SEVERE);
	LEVELS.emplace(SymbolTerm::intern("severe"), Level::SEVERE);
	}

PRED_loggable_1::StaticConstructor PRED_loggable_1::staticConstructor;

	PRED_loggable_1::PRED_loggable_1(Term *a0, Operation cont)
	{
		this->LARG[0] = a0; //0];
		this->cont = cont;
	}

	Operation PRED_loggable_1::exec(Prolog *engine) throw(PrologException)
	{

		Term *a1 = LARG[0]->dref();

		if (!(dynamic_cast<StructureTerm*>(a1) != nullptr) || a1->arity() != 2)
		{
			throw IllegalTypeException(this, 1, "package:level", a1);
		}

		Logger *logger = Logger::getLogger(a1->arg0(0)->name());
		if (!logger->isLoggable(LEVELS.getOrDefault(a1->arg0(1), Level::INFO)))
		{
			return engine->fail();
		}
		return cont;
	}
}
